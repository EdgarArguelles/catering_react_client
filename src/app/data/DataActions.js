/**
 * Following the Flux Standard Action: https://github.com/acdlite/flux-standard-action
 */
import keyMirror from 'keymirror';

export const ACTION_TYPES = keyMirror({
  DATA_CHANGE_VERSION: null,
});

export default class DataActions {
  static changeVersion(version) {
    return {
      type: ACTION_TYPES.DATA_CHANGE_VERSION,
      payload: {
        version,
      },
    };
  }
}