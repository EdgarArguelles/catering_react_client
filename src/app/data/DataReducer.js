import { createSlice } from '@reduxjs/toolkit';
import quotationsReducer from './quotations/QuotationsReducer';

const SLICE_NAME = 'DATA';

const dataSlice = createSlice({
  name: SLICE_NAME,
  initialState: {},
  extraReducers: builder => {
    return builder
      .addDefaultCase((state, action) => {
        state.quotations = quotationsReducer(state.quotations, action);
      });
  },
});

export default dataSlice.reducer;