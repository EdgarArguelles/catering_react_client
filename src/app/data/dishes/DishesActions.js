/**
 * Following the Flux Standard Action: https://github.com/acdlite/flux-standard-action
 */
import keyMirror from 'keymirror';
import Api from '../../common/Api';

export const ACTION_TYPES = keyMirror({
  FETCH_DISHES_REQUEST: null,
  FETCH_DISHES_SUCCESS: null,
  FETCH_DISHES_ERROR: null,
  FETCH_DISH_REQUEST: null,
  FETCH_DISH_SUCCESS: null,
  FETCH_DISH_ERROR: null,
});

const FIELDS = 'id name description picture price status categories{name}';

export default class DishesActions {
  static fetchDishes(courseTypeId) {
    const body = {query: `{courseType(id: ${courseTypeId}) {activeDishes{${FIELDS}}}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.FETCH_DISHES_REQUEST});
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.FETCH_DISHES_SUCCESS,
          payload: {
            courseTypeId,
            data: json.data.courseType.activeDishes,
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.FETCH_DISHES_ERROR,
          error: true,
          payload: error,
        });
        throw error;
      }
    };
  }

  static fetchDish(dishId) {
    const body = {query: `{dish(id: ${dishId}) {${FIELDS}}}`};

    return async dispatch => {
      try {
        dispatch({type: ACTION_TYPES.FETCH_DISH_REQUEST, payload: {dishId}});
        const json = await Api.graphql(dispatch, body);
        dispatch({
          type: ACTION_TYPES.FETCH_DISH_SUCCESS,
          payload: {
            dishId,
            data: json.data.dish,
          },
        });
        return json;
      } catch (error) {
        dispatch({
          type: ACTION_TYPES.FETCH_DISH_ERROR,
          error: true,
          payload: {dishId, error},
        });
        throw error;
      }
    };
  }
}