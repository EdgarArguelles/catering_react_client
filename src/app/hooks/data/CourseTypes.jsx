import {useDispatch, useSelector} from 'react-redux';
import {useQuery, useQueryClient} from 'react-query';
import Api from 'app/common/Api';

export const useCourseTypes = () => {
  const KEY = 'CourseTypes';
  const CACHE = 'courseTypesCached';
  const dispatch = useDispatch();
  const queryClient = useQueryClient();
  const isOnline = useSelector(state => state.app.isOnline);
  const cache = window.localStorage.getItem(CACHE);
  const courseTypesCached = cache ? JSON.parse(cache) : undefined;
  return useQuery(KEY, async () => {
    const body = {query: '{activeCourseTypes {id name picture position status}}'};
    const json = await Api.graphql(dispatch, body);
    const courseTypes = json.data.activeCourseTypes;
    window.localStorage.setItem(CACHE, JSON.stringify(courseTypes));
    return courseTypes;
  }, {
    retry: 5,
    retryDelay: 0,
    initialData: isOnline ? undefined : courseTypesCached,
    onError: () => {
      if (courseTypesCached) {
        queryClient.setQueryData(KEY, courseTypesCached);
      }
    },
  });
};

export const useDBVersion = courseTypes => {
  const KEY = 'DBVersion';
  const CACHE = 'versionCached';
  const dispatch = useDispatch();
  const dataVersion = parseInt(window.localStorage.getItem(CACHE), 10);
  return useQuery(KEY, async () => {
    const body = {query: '{version {version}}'};
    const json = await Api.graphql(dispatch, body);
    const version = json.data.version.version;
    if (version !== dataVersion) {
      window.localStorage.setItem(CACHE, version);
      // TODO: clean dishes from window.localStorage
    }
    return version;
  }, {enabled: !!courseTypes});
};