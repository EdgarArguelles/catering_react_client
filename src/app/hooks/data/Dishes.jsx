import {useDispatch, useSelector} from 'react-redux';
import {useQuery, useQueryClient} from 'react-query';
import Api from 'app/common/Api';

const DISH_KEY = 'Dish';
const CACHE = 'dishesCached';
const FIELDS = 'id name description picture price status categories{name}';

const getCache = () => {
  const cache = window.localStorage.getItem(CACHE);
  return cache ? JSON.parse(cache) : [];
};

const getActiveDishesCacheByCourseType = courseTypeId => {
  const cache = getCache();
  const filter = cache.filter(dish => dish.courseTypeId === courseTypeId && dish.status > 0);
  return filter.length ? filter : undefined;
};

const getDishCache = dishId => {
  const cache = getCache();
  const found = cache.find(dish => dish.id === dishId);
  return found ? found : undefined;
};

const addDishesCache = (courseTypeId, newDishes) => {
  const cache = getCache();
  const oldData = cache.filter(dish => dish.courseTypeId !== courseTypeId);
  window.localStorage.setItem(CACHE, JSON.stringify([...oldData, ...newDishes]));
};

const addDishCache = newDish => {
  const cache = getCache();
  const found = cache.find(dish => dish.id === newDish.id);
  !found && window.localStorage.setItem(CACHE, JSON.stringify([...cache, newDish]));
};

export const useActiveDishesByCourseType = courseTypeId => {
  const KEY = ['ActiveDishes', courseTypeId];
  const dispatch = useDispatch();
  const queryClient = useQueryClient();
  const isOnline = useSelector(state => state.app.isOnline);
  const setQueryData = () => {
    const cache = getActiveDishesCacheByCourseType(courseTypeId);
    cache && queryClient.setQueryData(KEY, cache);
  };

  return useQuery(KEY, async () => {
    const body = {query: `{courseType(id: ${courseTypeId}) {activeDishes{${FIELDS}}}}`};
    const json = await Api.graphql(dispatch, body);
    const dishes = json.data.courseType.activeDishes.map(dish => ({...dish, courseTypeId}));
    addDishesCache(courseTypeId, dishes);
    dishes.forEach(dish => queryClient.setQueryData([DISH_KEY, dish.id], dish));
    return dishes;
  }, {
    retry: 5,
    retryDelay: 0,
    initialData: isOnline ? undefined : getActiveDishesCacheByCourseType(courseTypeId),
    enabled: !!courseTypeId,
    onError: setQueryData,
  });
};

export const useDish = dishId => {
  const KEY = [DISH_KEY, dishId];
  const dispatch = useDispatch();
  const queryClient = useQueryClient();
  const isOnline = useSelector(state => state.app.isOnline);
  const setQueryData = () => {
    const cache = getDishCache(dishId);
    cache && queryClient.setQueryData(KEY, cache);
  };

  return useQuery(KEY, async () => {
    const body = {query: `{dish(id: ${dishId}) {${FIELDS}}}`};
    const json = await Api.graphql(dispatch, body);
    const dish = json.data.dish;
    addDishCache(dish);
    return dish;
  }, {
    retry: 5,
    retryDelay: 0,
    initialData: isOnline ? undefined : getDishCache(dishId),
    enabled: !!dishId,
    onError: setQueryData,
  });
};

/* export const useDishesByIds = dishesId => {
  // TODO: find how to call hook in loop
  dishesId.forEach(dishId => {
    useDish(dishId);
  });
}; */