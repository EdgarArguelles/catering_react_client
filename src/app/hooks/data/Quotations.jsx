import {useDispatch, useSelector} from 'react-redux';
import {useInfiniteQuery} from 'react-query';
import Api from 'app/common/Api';
import Utils from 'app/common/Utils';

export const useQuotations = () => {
  const KEY = 'Quotations';
  const dispatch = useDispatch();
  const loggedUser = useSelector(state => state.auth.loggedUser);
  const DEFAULT_PAGE_PARAM = {page: 0, size: 5, sort: ['createdAt'], direction: 'DESC'};
  const results = useInfiniteQuery(KEY, async ({pageParam = DEFAULT_PAGE_PARAM}) => {
    const FIELDS = 'totalElements totalPages content{id name createdAt price}';
    const body = {query: `{quotationPage(pageDataRequest: ${Utils.stringifyPageDataRequest(pageParam)}) {${FIELDS}}}`};
    const json = await Api.graphql(dispatch, body);
    const {content, totalElements, totalPages} = json.data.quotationPage;
    return {
      quotations: content,
      metaData: {pagination: {...pageParam}, totalElements, totalPages},
    };
  }, {
    enabled: !!loggedUser,
    getNextPageParam: ({metaData: {pagination, totalPages}}) =>
      pagination.page + 1 < totalPages ? {...pagination, page: pagination.page + 1} : undefined,
  });
  const quotations = results.data?.pages.map((page => page.quotations)).flat();
  const metaData = results.data?.pages.map((page => page.metaData)).pop();
  return {...results, quotations, metaData};
};