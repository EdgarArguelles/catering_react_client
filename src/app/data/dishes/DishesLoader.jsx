import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {useDispatch, useSelector} from 'react-redux';
import {areAllDishesPresent, fetchDishesList} from 'app/features/quotations/dish/Dish.service';
import DishesActions from './DishesActions';

const DishesLoader = ({children, renderer, loader, dishes}) => {
  const dispatch = useDispatch();
  const dishFetching = useSelector(state => state.data.fetching.dish);
  const allDishes = useSelector(state => state.data.dishes);

  useEffect(() => {
    const fetchDish = dishId => dispatch(DishesActions.fetchDish(dishId));

    fetchDishesList(dishes, allDishes, dishFetching, fetchDish);
  }, [dishes, dishFetching, allDishes, dispatch]);

  if (areAllDishesPresent(dishes, allDishes) || dishes.length === 0) {
    if (renderer) {
      return renderer();
    }

    return <>{children}</>;
  }

  if (loader) {
    return loader;
  }

  return null;
};

DishesLoader.propTypes = {
  children: PropTypes.object,
  renderer: PropTypes.func,
  loader: PropTypes.object,
  dishes: PropTypes.array.isRequired,
};

export default React.memo(DishesLoader);