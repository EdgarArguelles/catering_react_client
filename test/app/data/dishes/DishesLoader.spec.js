/* eslint-disable max-lines */
import sinon from 'sinon';
import renderer, {act} from 'react-test-renderer';
import React from 'react';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {createStore} from 'redux';
import reducers from 'app/Reducers';
import DishesLoader from 'app/data/dishes/DishesLoader';
import DishesActions, {ACTION_TYPES} from 'app/data/dishes/DishesActions';
import {ACTION_TYPES as DATA_TYPES} from 'app/data/DataActions';

describe('Data -> Dishes -> Loader', () => {
  const myRenderer = () => <div>renderer</div>;
  const dispatchStub = sinon.stub();
  const fetchDishStub = sinon.stub(DishesActions, 'fetchDish');
  let component, store, wrapper;

  afterEach(() => {
    dispatchStub.reset();
    fetchDishStub.reset();
  });

  const mountComponent = (state, props, children, useRealStore) => {
    store = configureStore()(state);
    store.dispatch = dispatchStub;
    store = useRealStore ? createStore(reducers, state) : store;
    wrapper = renderer.create(<Provider store={store}>
      <DishesLoader {...props}>{children}</DishesLoader>
    </Provider>);
    component = wrapper.root.find(el => el.type.name === 'DishesLoader');
  };

  describe('load', () => {
    it('should have required props', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {}},
          },
        },
        {
          dishes: [{id: 'D1'}, {id: 'D2'}, {id: 'D3'}],
        },
      );

      expect(component.props.children).toBeUndefined();
      expect(component.props.renderer).toBeUndefined();
      expect(component.props.loader).toBeUndefined();
      expect(component.props.dishes).toStrictEqual([{id: 'D1'}, {id: 'D2'}, {id: 'D3'}]);
    });

    it('should have all props', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {D1: false, D2: true}},
            dishes: {D1: {id: 'D1'}, D2: {id: 'D2'}, D3: {id: 'D3'}, D4: {id: 'D4'}},
          },
        },
        {
          renderer: myRenderer,
          loader: <div>loader</div>,
          dishes: [{id: 'D1'}, {id: 'D2'}, {id: 'D3'}, {id: 'D4'}],
        },
        <h1>Test</h1>,
      );

      expect(component.props.children).toStrictEqual(<h1>Test</h1>);
      expect(component.props.renderer).toStrictEqual(myRenderer);
      expect(component.props.loader).toStrictEqual(<div>loader</div>);
      expect(component.props.dishes).toStrictEqual([{id: 'D1'}, {id: 'D2'}, {id: 'D3'}, {id: 'D4'}]);
    });
  });

  describe('useEffect -> fetchDish', () => {
    beforeEach(() => {
      fetchDishStub.withArgs('D1').returns({type: 'TEST'});
      fetchDishStub.withArgs('D2').returns({type: 'TEST'});
      fetchDishStub.withArgs('D3').returns({type: 'TEST'});
      fetchDishStub.withArgs('D4').returns({type: 'TEST'});
      mountComponent(
        {
          data: {
            fetching: {dish: {D1: true}},
            dishes: {D2: {id: 'D2'}, D3: {id: 'D3'}, D4: {id: 'D4'}},
          },
        },
        {
          dishes: [{id: 'D1'}, {id: 'D2'}, {id: 'D3'}, {id: 'D4'}],
        },
        null, true,
      );
    });

    it('should not call fetchDish', () => {
      sinon.assert.callCount(fetchDishStub, 0);
      sinon.assert.callCount(dispatchStub, 0);
    });

    it('should call fetchDish', () => {
      act(() => {
        store.dispatch({type: ACTION_TYPES.FETCH_DISH_ERROR, payload: {dishId: 'D1'}});
        store.dispatch({
          type: DATA_TYPES.DATA_CHANGE_VERSION,
          payload: {
            version: 6,
          },
        });
      });

      sinon.assert.callCount(fetchDishStub, 4);
      sinon.assert.calledWithExactly(fetchDishStub, 'D1');
      sinon.assert.calledWithExactly(fetchDishStub, 'D2');
      sinon.assert.calledWithExactly(fetchDishStub, 'D3');
      sinon.assert.calledWithExactly(fetchDishStub, 'D4');
      sinon.assert.callCount(dispatchStub, 0);
    });
  });

  describe('snapshot', () => {
    it('should render renderer', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {}},
            dishes: {D1: {id: 'D1'}, D2: {id: 'D2'}, D3: {id: 'D3'}},
          },
        },
        {
          renderer: myRenderer,
          loader: <div>loader</div>,
          dishes: [{id: 'D1'}, {id: 'D2'}],
        },
        <h1>Test</h1>,
      );

      global.expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render h1 Test', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {}},
            dishes: {},
          },
        },
        {
          loader: <div>loader</div>,
          dishes: [],
        },
        <h1>Test</h1>,
      );

      global.expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render loader', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {}},
            dishes: {D1: {id: 'D1'}},
          },
        },
        {
          renderer: myRenderer,
          loader: <div>loader</div>,
          dishes: [{id: 'D1'}, {id: 'D2'}, {id: 'D3'}, {id: 'D4'}],
        },
        <h1>Test</h1>,
      );

      global.expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render null', () => {
      mountComponent(
        {
          data: {
            fetching: {dish: {}},
          },
        },
        {
          renderer: myRenderer,
          dishes: [{id: 'D1'}],
        },
        <h1>Test</h1>,
      );

      global.expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });
});