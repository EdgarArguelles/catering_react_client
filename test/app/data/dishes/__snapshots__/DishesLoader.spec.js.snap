// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data -> Dishes -> Loader snapshot should render h1 Test 1`] = `
<h1>
  Test
</h1>
`;

exports[`Data -> Dishes -> Loader snapshot should render loader 1`] = `
<div>
  loader
</div>
`;

exports[`Data -> Dishes -> Loader snapshot should render null 1`] = `null`;

exports[`Data -> Dishes -> Loader snapshot should render renderer 1`] = `
<div>
  renderer
</div>
`;
